### PR TITLE
Issue 369: Auto Freeze Phase 3

### PR DESCRIPTION
**Pull request by vkumra-broad**
Auto Freeze Phase 3

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorSupervisor.scala [hunk 1]
```diff
@@ -110,6 +121,18 @@ class ClusterMonitorSupervisor(monitorConfig: MonitorConfig, dataprocConfig: Dat
     }
   }
 
+  def autoFreezeClusters: Future[Unit] = {
+    val clusterList = dbRef.inTransaction {
+      _.clusterQuery.getClusterReadyToAutoFreeze(autoFreezeConfig.autoFreezeAfter)
+    }
+    clusterList map { cl =>
+      cl.foreach(c => {
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorSupervisor.scala]: Nitpick: don't need the {

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorSupervisor.scala [hunk 2]
```diff
@@ -110,6 +121,18 @@ class ClusterMonitorSupervisor(monitorConfig: MonitorConfig, dataprocConfig: Dat
...
+        logger.info(s"Auto freezing cluster ${c.clusterName} in project ${c.googleProject}")
+        leoService.internalStopCluster(c)
```
[the remark this answers is not in the document]
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorSupervisor.scala]: Maybe make it a warning not an error though

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponentSpec.scala [hunk 3]
```diff
@@ -398,4 +400,63 @@ class ClusterComponentSpec extends TestComponent with FlatSpecLike with CommonTe
...
+    dbFutureValue { _.clusterQuery.save(c1, gcsPath("gs://bucket1"), Some(serviceAccountKey.id)) } shouldEqual c1
+    dbFutureValue { _.clusterQuery.save(c2, gcsPath("gs://bucket1"), Some(serviceAccountKey.id)) } shouldEqual c2
+
+    dbFutureValue { _.clusterQuery.getClustersReadyToAutoFreeze(autoFreezeconfig.autoFreezeAfter) } shouldBe List.empty
+
+    eventually(timeout(Span(30, Seconds))) {
+      val autoFreezeList = dbFutureValue { _.clusterQuery.getClustersReadyToAutoFreeze(autoFreezeconfig.autoFreezeAfter) }
+      autoFreezeList should contain (c1)
+      //c2 is already stopped
+      autoFreezeList should not contain c2
```
Comment (kyuksel) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponentSpec.scala]: Ideally, we would also test the other stoppable and unstoppable statuses as well. I'm wary of creating more duplicate code with `c3`, `c4`, etc., though as they caused a lot of code updates when I was updating the `Cluster` `case class`. It'd be great to have a `def getCluster: Cluster` (perhaps in `CommonTestData`) that yields the most commonly used configuration and every other `cX` would do something like 

```scala
val c2 = getCluster.copy(name = "cluster2", status = "Updating", etc.)
```

I know this is not something introduced by this PR and shouldn't be a blocker but I'd be happy to add this if you like...just let me know.

**Comment by kyuksel** [comment]
Good work. I've added some comments and questions inline.

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponentSpec.scala (hunk 3 above)
Comment (vkumra-broad) [reply src/test/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponentSpec.scala]: +1 for this. I'll create a housekeeping ticket for this.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorSupervisor.scala [hunk 4]
```diff
@@ -110,6 +121,20 @@ class ClusterMonitorSupervisor(monitorConfig: MonitorConfig, dataprocConfig: Dat
...
 
+  def autoFreezeClusters: Future[Unit] = {
+    val clusterList = dbRef.inTransaction {
+      _.clusterQuery.getClustersReadyToAutoFreeze(autoFreezeConfig.autoFreezeAfter)
+    }
+    clusterList map { cl =>
+      cl.foreach { c =>
+        logger.info(s"Auto freezing cluster ${c.clusterName} in project ${c.googleProject}")
+        leoService.internalStopCluster(c).failed.foreach { e =>
+          logger.warn(s"Error occurred auto freezing cluster ${c.projectNameString}", e)
```
[the remark this answers is not in the document]
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorSupervisor.scala]: Well we log a warning, which will be picked up by sentry/kibana. Maybe we could configure a sentry alert for this so at least we get notified?

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponentSpec.scala (hunk 3 above)
Comment (rtitle) [reply src/test/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponentSpec.scala]: +1 on creating common cluster instances/helper methods that can be reused across tests. There is currently a lot of repeated boilerplate in test setup code.

Comment (vkumra-broad) [reply src/test/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponentSpec.scala]: https://github.com/DataBiosphere/leonardo/issues/394

File: src/main/resources/reference.conf [hunk 5]
```diff
@@ -90,7 +90,8 @@ serviceAccounts {
 }
 
 autoFreeze {
+  enableAutoFreeze = true
```
Comment (rtitle) [inline src/main/resources/reference.conf]: Make this false (maybe with a comment) so we can merge this

**Pull request merged**